pkg/agent/controller/traceflow/traceflow_controller.go

```diff
@@ -295,10 +325,16 @@ func (c *Controller) injectPacket(tf *opsv1alpha1.Traceflow) error {
 			dstIP = dstPod.Status.PodIP
 			dstNodeIP = dstPod.Status.HostIP
 		}
+	} else if tf.Spec.Destination.Service != "" {
+		dstSvc, err := c.serviceLister.Services(tf.Spec.Destination.Namespace).Get(tf.Spec.Destination.Service)
+		if err != nil {
+			return err
+		}
+		dstIP = dstSvc.Spec.ClusterIP
 	}
-	if dstNodeIP != "" {
+	if dstNodeIP != "" || tf.Spec.Destination.Pod == "" {
 		peerIP := net.ParseIP(dstNodeIP)
-		if c.networkConfig.TunnelType == ovsconfig.GeneveTunnel && peerIP != nil && c.networkConfig.TrafficEncapMode.NeedsEncapToPeer(peerIP, c.nodeConfig.NodeIPAddr) {
+		if c.networkConfig.TunnelType == ovsconfig.GeneveTunnel && (tf.Spec.Destination.Pod == "" || c.networkConfig.TrafficEncapMode.NeedsEncapToPeer(peerIP, c.nodeConfig.NodeIPAddr)) {
```

Comment:
I'm getting a bit confused by these if statements, so I think some comments may be useful here.

I didn't spend too much time deciphering this, but for the `tf.Spec.Destination.IP != ""` case, shouldn't we test if the IP is a local IP? For example, if `tf.Spec.Destination.IP` is a local Pod IP, then we don't need this check. Or is it too much of a corner case? In any case, comments would be good IMO.

Reply:
Added comments here.
The tf.Spec.Destination.IP != "" case is enhanced in PR #981